File: apps/web/src/types/creative.ts
export interface CreativePolicy {
  roas_floor: number;
  warn_threshold: number;
  block_threshold: number;
  min_impressions: number;
}

export interface CreativeSummary {
  creative_count: number;
  active_creatives: number;
  blocked_creatives: number;
  watchlist_creatives: number;
  average_roas: number;
  median_roas: number;
  active_spend_share: number;
  watchlist_spend_share: number;
  blocked_spend_share: number;
  guardrail_counts: Record<string, number>;
}

export interface CreativeHighlight {
  creative_id: string;
  channel: string;
  roas_adjusted: number;
  brand_safety_score: number;
  status: string;
}

export interface CreativeRow {
  creative_id: string;
  channel: string;
  impressions: number;
  clicks: number;
  conversions: number;
  spend: number;
  revenue: number;
  brand_safety_score: number;
  brand_safety_tier: string;
  brand_safety_factor: number;
  sample_size_factor: number;
  ctr: number;
  cvr: number;
  aov: number;
  roas_smoothed: number;
  roas_adjusted: number;
  guardrail_factor: number;
  status: string;
  guardrail: string | null;
  spend_share: number;
  profit_expectation: number;
}

export interface CreativeChannelGuardrail {
  channel: string;
  creative_count: number;
  active_creatives: number;
  watchlist_creatives: number;
  blocked_creatives: number;
  flagged_creatives: number;
  active_spend_share: number;
  watchlist_spend_share: number;
  blocked_spend_share: number;
  flagged_spend_share: number;
  average_roas: number;
  average_brand_safety: number;
  top_guardrail: string | null;
  top_guardrail_count: number;
  representative_creative: string | null;
  representative_status: string | null;
}

export interface CreativeResponseReport {
  generated_at: string;
  policy: CreativePolicy;
  summary: CreativeSummary;
  top_creatives: CreativeHighlight[];
  creatives: CreativeRow[];
  channel_guardrails: CreativeChannelGuardrail[];
}

File: shared/schemas/creative.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class CreativePolicy(BaseModel):
    roas_floor: float
    warn_threshold: float
    block_threshold: float
    min_impressions: int


class CreativeSummary(BaseModel):
    creative_count: int
    active_creatives: int
    blocked_creatives: int
    watchlist_creatives: int
    average_roas: float
    median_roas: float
    active_spend_share: float
    watchlist_spend_share: float
    blocked_spend_share: float
    guardrail_counts: dict[str, int] = Field(default_factory=dict)


class CreativeHighlight(BaseModel):
    creative_id: str
    channel: str
    roas_adjusted: float
    brand_safety_score: float
    status: str


class CreativeRow(BaseModel):
    creative_id: str
    channel: str
    impressions: int
    clicks: int
    conversions: int
    spend: float
    revenue: float
    brand_safety_score: float
    brand_safety_tier: str
    brand_safety_factor: float
    sample_size_factor: float
    ctr: float
    cvr: float
    aov: float
    roas_smoothed: float
    roas_adjusted: float
    guardrail_factor: float
    status: str
    guardrail: str | None
    spend_share: float
    profit_expectation: float


class CreativeResponseReport(BaseModel):
    generated_at: datetime = Field(default_factory=datetime.utcnow)
    policy: CreativePolicy
    summary: CreativeSummary
    top_creatives: list[CreativeHighlight]
    creatives: list[CreativeRow]


__all__ = [
    "CreativePolicy",
    "CreativeResponseReport",
    "CreativeRow",
    "CreativeSummary",
    "CreativeHighlight",
]
